src/NLog.Targets.Syslog/MessageStorage/MultiEncodingStreamWriter.cs

```diff
@@ -0,0 +1,40 @@
+// Licensed under the BSD license
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace NLog.Targets.Syslog.MessageStorage
+{
+    internal class MultiEncodingStreamWriter : IDisposable
+    {
+        private readonly Dictionary<Type, StreamWriter> streamWriterForEncodingType;
+
+        public MultiEncodingStreamWriter(Stream stream)
+        {
+            var ascii = new StreamWriter(stream, new ASCIIEncoding()) { AutoFlush = true };
+            var utf8 = new StreamWriter(stream, new UTF8Encoding(false)) { AutoFlush = true };
+
+            streamWriterForEncodingType = new Dictionary<Type, StreamWriter>
+            {
+                { typeof(ASCIIEncoding), ascii },
+                { Encoding.ASCII.GetType(), ascii },
+                { typeof(UTF8Encoding), utf8 },
+                { Encoding.UTF8.GetType(), utf8 }
+            };
+        }
+
+        public void Write(Encoding encoding, string s)
+        {
+            streamWriterForEncodingType[encoding.GetType()].Write(s);
```

Comment:
Strange to perform GetType()-call and dictionary-lookup for every string-write. Any reason why having both Asciii AND UTF8 ? Why not use UTF8 for everything?

Reply:
If RFC3164 is used and a client application sends UTF8 chars they should be encoded as the ASCII question mark

The dictionary lookup doesn't have an impact being O(1) and GetType [looks](https://stackoverflow.com/a/353435) pretty fast

I will try to improve this a bit in the future, but I am not a fan of premature micro-optimizations if not supported by a profiler report or benchmarks